gramineproject/gramine
# [Pal/Linux-SGX] Implement gramine-sgx-get-token in Rust

## Description of the changes 

This PR migrates our implementation of `gramine-sgx-get-token` to Rust. This is because:
- The tool is small and self-contained, which makes it a good test case for figuring out how to include Rust code in the project and build it with Meson.
- Future plans involve having `gramine-sgx` generate the token when the enclave is first launched, which requires an implementation we can link to Gramine itself. Python is not suitable for that.

While Meson has some built-in support for building Rust code, it [does not support making use of any crates](https://github.com/mesonbuild/meson/issues/2173), instead insisting on building each dependency with Meson. This is so impractical that I use `custom_target` instead.

One question that might come up is the usage of dependencies from crates.io.
- One approach would be to reject them entirely and attempt to write everything ourselves. I don't think that is reasonable. In my opinion, some crates in the ecosystem are used commonly enough that they are an integral part of Rust's value offering (e.g. `serde`). Indeed, external crates are considered easily accessible and are often cited as reasons to avoid additions to `std` (e.g. `rand`).
- As far as source code security goes, `Cargo.lock` contains the source tarball hashes of all dependencies used. This is already no worse than some other dependencies Gramine downloads during `meson setup`.
  - Of course there needs to be some way of vetting the code when doing `cargo update`. [`cargo crev`](https://github.com/crev-dev/cargo-crev) is one possible solution I'm aware of, which involves a web of trust with users signing "I have reviewed the code and haven't found anything suspicious" or "I am the author of this crate". 
  - Otherwise we're trusting the [crates.io registry](https://github.com/rust-lang/crates.io-index/) and the authors of the crates to keep their auth safe. This is not perfect, but also not that bad.
- Apart from the generic crates, some might be worth special attention in that they have been published as part of Fortanix's Rust EDP (currently the code uses `aesm_client`, `sgxs`, and `sgx_isa`). Like everything else though, they are licensed appropriately and work for our usecase. In the worst case, we can always fork these. 

## How to test this PR? 

I have done the following:

- Check that meson/ninja appropriately rebuild the Rust code when it is modified.
- Verify that the token gets generated correctly for both `oot` setups (where AESM is required) and non-`oot` setups (where a dummy token is used).
- Check that the optional CPU features get added to the XFRM. I did not have the opportunity to test every single one, but the rest should work if the previous Python code worked.
